File: leetcode/1342.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Number of Steps to Reduce a Number to Zero.
// Memory Usage: 5.8 MB, less than 95.31% of C++ online submissions for Number of Steps to Reduce a Number to Zero.
class Solution {
 public:
  int numberOfSteps(int num) {
    int res = 0;
    while (num != 0) {
      if (num % 2 == 1) {
        num -= 1;
      } else {
        num /= 2;
      }
      res += 1;
    }
    return res;
  }
};

File: leetcode/1342.py
"""
Runtime: 37 ms, faster than 43.28% of Python3 online submissions for Number of Steps to Reduce a Number to Zero.
Memory Usage: 13.9 MB, less than 88.59% of Python3 online submissions for Number of Steps to Reduce a Number to Zero.
"""
# bit
class Solution:
    def numberOfSteps(self, num: int) -> int:
        # binary is like 0000101001
        #                xxxxssssss
        # the answer will be num of s - 1 (for divides)
        # and we need to add how many those are 1s
        res = -1
        while num:
            res += (num & 0x00000001) + 1
            num >>= 1
        return max(res, 0)


"""
Runtime: 24 ms, faster than 95.97% of Python3 online submissions for Number of Steps to Reduce a Number to Zero.
Memory Usage: 14.2 MB, less than 68.74% of Python3 online submissions for Number of Steps to Reduce a Number to Zero.
"""


class Solution:
    def numberOfSteps(self, num: int) -> int:
        res = 0
        while num:
            if num & 1:
                num -= 1
            else:
                num >>= 1
            res += 1

        return res
